snuba/datasets/cdc/groupedmessage_processor.py

```diff
@@ -1,54 +1,76 @@
+from datetime import datetime
+from typing import Any, Mapping, Optional
+
+from dataclasses import dataclass
 from dateutil.parser import parse as dateutil_parse
-from snuba.datasets.cdc.cdcprocessors import CdcProcessor
+from snuba.datasets.cdc.cdcprocessors import CdcProcessor, CDCMessageRow
 
 
-class GroupedMessageProcessor(CdcProcessor):
+@dataclass(frozen=True)
+class GroupedMessageRow(CDCMessageRow):
+    offset: int
+    id: int
+    record_deleted: bool
+    status: Optional[int] = None
+    last_seen: Optional[datetime] = None
+    first_seen: Optional[datetime] = None
+    active_at: Optional[datetime] = None
+    first_release_id: Optional[int] = None
 
-    def __init__(self, postgres_table):
-        super(GroupedMessageProcessor, self).__init__(
-            pg_table=postgres_table,
+    @classmethod
+    def from_wal(cls, offset, columnnames, columnvalues):
+        raw_data = dict(zip(columnnames, columnvalues))
+        return GroupedMessageRow(
+            offset=offset,
+            id=raw_data['id'],
+            record_deleted=False,
+            status=raw_data['status'],
+            last_seen=dateutil_parse(raw_data['last_seen']),
+            first_seen=dateutil_parse(raw_data['first_seen']),
+            active_at=dateutil_parse(raw_data['active_at']),
+            first_release_id=raw_data['first_release_id'],
         )
 
-    def __build_record(self, offset, columnnames, columnvalues):
-        raw_data = dict(zip(columnnames, columnvalues))
-        output = {
-            'offset': offset,
-            'id': raw_data['id'],
-            'record_deleted': 0,
-            'status': raw_data['status'],
-            'last_seen': dateutil_parse(raw_data['last_seen']),
-            'first_seen': dateutil_parse(raw_data['first_seen']),
-            'active_at': dateutil_parse(raw_data['active_at']),
-            'first_release_id': raw_data['first_release_id'],
+    @classmethod
+    def from_bulk(cls, row):
+        return GroupedMessageRow(
+            offset=0,
```

Comment:
Shouldn't this be `None`? Even if the ClickHouse value isn't nullable, shouldn't that transformation happen in`to_clickhouse` and not here?

Reply:
This is a conventional value for the bulk updates that just has to be smaller than any offset it will be consumed later. I am ok using None.

Reply:
Actually thinking over it. In order to make this None, it means the type has to be optional for everyone. I think it may be worse that way.

Reply:
Isn't that true though (at least conceptually, if not literally), at least for anybody accessing `GroupedMessageRow` directly? Are there other classes doing direct attribute reference on these instances, other than `to_clickhouse` which is already retyping it as `Any` already?